File: leetcode/0001-1000/0901-1000/0933-number-of-recent-calls/NumberOfRecentCalls.java
import java.util.ArrayDeque;
import java.util.Deque;

public class NumberOfRecentCalls {
  public static void main(String[] args) {
    //
  }
}

class RecentCounter {
  Deque<Integer> queue;

  public RecentCounter() {
    queue = new ArrayDeque<>();
  }

  public int ping(int t) {
    while (!queue.isEmpty() && queue.peekFirst() < t - 3000) {
      queue.pollFirst();
    }

    queue.offerLast(t);

    return queue.size();
  }
}

/**
 * Your RecentCounter object will be instantiated and called as such:
 * RecentCounter obj = new RecentCounter();
 * int param_1 = obj.ping(t);
 */

File: leetcode/0001-1000/0901-1000/0933-number-of-recent-calls/NumberOfRecentCalls.ts
class RecentCounter {
  queue: number[];
  constructor() {
    this.queue = []
  }

  ping(t: number): number {
    while (this.queue.length > 0 && this.queue[0] < t - 3000) {
      this.queue.shift();
    }

    this.queue.push(t);

    return this.queue.length;
  }
}

